src/compositemark/boxplot.ts

```diff
@@ -165,72 +153,71 @@ export function normalizeBoxPlot(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT
   };
 }
 
-export function box2DOrient(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT | BoxPlotDef>): Orient {
+export function boxOrient(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT | BoxPlotDef>): Orient {
   const {mark: mark, encoding: encoding, ...outerSpec} = spec;
 
-  // FIXME: refactor code such that we don't have to do this casting
-  // We can cast here as we already check from outside that both x and y are FieldDef
-  const xDef = encoding.x as FieldDef<Field>;
-  const yDef = encoding.y as FieldDef<Field>;
-  let resultOrient: Orient;
+  const continuousXAxis = isFieldDef(encoding.x) && isContinuous(encoding.x);
+  const continuousYAxis = isFieldDef(encoding.y) && isContinuous(encoding.y);
 
-  if (isDiscrete(xDef) && isContinuous(yDef)) {
-    resultOrient = 'vertical';
-  } else if (isDiscrete(yDef) && isContinuous(xDef)) {
-    resultOrient = 'horizontal';
-  } else {
-    if (isContinuous(xDef) && isContinuous(yDef)) {
-      if (xDef.aggregate === undefined && yDef.aggregate === BOXPLOT) {
-        resultOrient = 'vertical';
-      } else if (yDef.aggregate === undefined && xDef.aggregate === BOXPLOT) {
-        resultOrient = 'horizontal';
-      } else if (xDef.aggregate === BOXPLOT && yDef.aggregate === BOXPLOT) {
-        throw new Error('Both x and y cannot have aggregate');
-      } else {
-        if (isBoxPlotDef(mark)) {
-          if (mark && mark.orient) {
-            resultOrient = mark.orient;
-          } else {
-            // default orientation = vertical
-            resultOrient = 'vertical';
-          }
-        } else {
-          resultOrient = 'vertical';
-        }
-      }
+  if (continuousXAxis
+      && (encoding.y === undefined || (isFieldDef(encoding.y) && isDiscrete(encoding.y)))) {
+    return 'horizontal';
+  } else if (continuousYAxis
+      && (encoding.x === undefined || (isFieldDef(encoding.x) && isDiscrete(encoding.x)))) {
+    return 'vertical';
+  }
+
+  if (!continuousXAxis && !continuousYAxis) {
+    throw new Error('Need a valid continuous axis for boxplots');
+  }
+
+  if (isFieldDef(encoding.x) && isFieldDef(encoding.y) && isContinuous(encoding.x) && isContinuous(encoding.y)) {
```

Comment:
@kanitw Only reason I didn't use continuousXAxis and continuousYAxis is because of that typeguard issue I asked you about. encoding.x and encoding.y aren't typed as FieldDef's and I need to access their aggregate property inside these if blocks